core/src/main/java/tc/oc/pgm/filters/dynamic/FilterMatchModule.java

```diff
@@ -0,0 +1,605 @@
+    this.listeners
+        .columnMap()
+        .forEach(
+            (scope, column) -> {
+              if (scope.isInstance(filterable)) {
+                // For each filter in this scope
+                column.forEach(
+                    (filter, filterListeners) -> {
+                      final Boolean before;
+                      final boolean after;
+                      if (beforeCache.containsKey(filter)) {
+                        // If the filter has already been checked, we have both responses saved.
+                        before = beforeCache.get(filter);
+                        after = afterCache.get(filter);
+                      } else {
+                        // The first time a particular filter is checked, move the old response to
+                        // a local temporary cache and save the new response to the permanent cache.
+                        before = afterCache.get(filter);
+                        beforeCache.put(filter, before);
+                        after = filter.response(query);
+                        afterCache.put(filter, after);
+                      }
+
+                      if (before == null || before != after) {
+                        dispatches.add(
+                            () ->
+                                (after ? filterListeners.rise : filterListeners.fall)
+                                    .forEach(
+                                        listener ->
+                                            dispatch(
+                                                (FilterListener<? super F>) listener,
+                                                filter,
+                                                filterable,
+                                                after)));
+                      }
+                    });
+              }
+            });
+  }
+
+  /**
+   * Checks the response for a query for all filters that fits a scope, if the response is different
+   * than the last cached response and the filters cares about the change the responses are
+   * dispatched to all the filters after the check is done.
+   *
+   * @param filterable the scope for this check
+   * @param query the query to check against some filters
+   */
+  private <F extends Filterable<?>, Q extends Query> void check(F filterable, Q query) {
+    final List<Runnable> dispatches = new ArrayList<>();
+    check(filterable, query, dispatches);
+    dispatches.forEach(Runnable::run);
+  }
+
+  @Override
+  public void tick(Match match, Tick tick) {
+    if (!match.isRunning()) return;
+
+    this.tick();
+  }
+
+  public void tick() {
+    final Set<Filterable<?>> checked = new HashSet<>();
+    for (; ; ) {
+      // Collect Filterables that are dirty, and have not already been checked in this tick
+      final Set<Filterable<?>> checking = ImmutableSet.copyOf(Sets.difference(dirtySet, checked));
+      if (checking.isEmpty()) break;
```

Comment:
```suggestion
    final Set<Filterable<?>> checked = new HashSet<>(), checking;
    while (!(checking = ImmutableSet.copyOf(Sets.difference(dirtySet, checked))).isEmpty()) {
```
It may not look the cleanest, but at least it's not an infinite loop and the exit condition is clear
Could probably put `ImmutableSet.copyOf(Sets.difference(dirtySet, checked))` in a function and call that for clarity:
`while (!(checking = getMissingChecks()).isEmpty())`